### PR TITLE
Mock TRC

### PR DESCRIPTION
**Pull request by ieperen3039**
Dummy TRC

### CODE REVIEW

File: cmd/trcd/main.go [hunk 1]
```diff
@@ -29,31 +33,89 @@ func init() {
...
+					case <-closeCh:
+						return
+					}
+				}
+			}()
+		}
 	}
 }
+
+func setStateHandler(msg *api.Message) (*api.Message, error) {
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: can you make those anonymous or, (as a compromise) define them in the function itself, like so:

File: cmd/trcd/main.go [hunk 2]
```diff
@@ -29,31 +33,89 @@ func init() {
...
+		}
+
+		// TODO initial state
+		if err := trcConn.SendState(&api.State{}); err != nil {
+			log.Error(err)
 			return
 		}
+
+		closeCh := make(chan struct{})
+		// TODO close when trc should shutdown
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: capitalize `TRC`

File: cmd/trcd/main.go [hunk 3]
```diff
@@ -29,31 +33,89 @@ func init() {
...
-				log.Printf("Failed to decode input: %s", err)
+			for err := range trcConn.Errors() {
+				log.Error(err)
 			}
-			log.Printf("Got: %v", msg)
 		}()
-		if _, err := io.Copy(c, os.Stdin); err != nil {
-			log.Printf("Failed to copy input: %s", err)
+
+		// mock trc is running parallel to client anyway
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: what's the purpose of this comment?

File: cmd/trcd/main.go [hunk 4]
```diff
@@ -29,31 +33,89 @@ func init() {
...
 		}
-		defer c.Close()
+		defer sockConn.Close()
 
+		trcConn := Connect(sockConn, sockConn,
+			WithHandler(api.MessageTypeState, setStateHandler),
+			WithHandler(api.MessageTypePing, pingHandler),
+		)
+
+		// print errors resulting from mock trc
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: don't need to explain this.

File: cmd/trcd/main.go [hunk 5]
```diff
@@ -29,31 +33,89 @@ func init() {
...
+		if err := trcConn.SendState(&api.State{}); err != nil {
+			log.Error(err)
 			return
 		}
+
+		closeCh := make(chan struct{})
+		// TODO close when trc should shutdown
+		defer close(closeCh)
+
+		// send random messages unless specified
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: unless what is specified?

File: cmd/trcd/main.go [hunk 6]
```diff
@@ -29,31 +33,89 @@ func init() {
 func main() {
 	flag.Parse()
 
-	conn, err := net.Listen("unix", *sock)
+	var netConn net.Listener
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: this is not a `Conn`, it's a `Listener`, `netLst` maybe?

File: cmd/trcd/main.go [hunk 7]
```diff
@@ -1,23 +1,27 @@
...
 var (
-	sock = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	sock   = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	useTCP = flag.Bool("tcp", false, "Use a tcp socket instead of a Unix socket")
+	silent = flag.Bool("silent", false, "Disables automatic sending of random state updates")
 )
 
 func init() {
-	log.SetLevel(log.InfoLevel)
+	log.SetLevel(log.DebugLevel)
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: Add a flag like it's done in `srrs`, don't hard-code debug logging.

File: cmd/trcd/main.go [hunk 8]
```diff
@@ -1,23 +1,27 @@
...
 	"github.com/rvolosatovs/turtlitto/pkg/api"
+	"github.com/rvolosatovs/turtlitto/pkg/trcapi"
+	. "github.com/rvolosatovs/turtlitto/pkg/trcapi/trctest"
 )
 
 var (
-	sock = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	sock   = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	useTCP = flag.Bool("tcp", false, "Use a tcp socket instead of a Unix socket")
+	silent = flag.Bool("silent", false, "Disables automatic sending of random state updates")
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: Why would we want this?

File: cmd/srrs/main_test.go [hunk 9]
```diff
@@ -110,24 +111,51 @@ func TestAll(t *testing.T) {
...
+			a.NoError(err)
+			log.Debug("Random command sent")
+
+			got := &api.State{}
+			log.Debug("Receiving random state on WebSocket...")
+			err = wsConn.ReadJSON(got)
+			a.Nil(err)
+			a.Equal(got.Command, state.Command)
+			log.Debug("Random state received on WebSocket")
+		})
```
Comment (rvolosatovs) [inline cmd/srrs/main_test.go]: You should send commands/states via web API here

File: cmd/srrs/main.go [hunk 10]
```diff
@@ -26,6 +26,7 @@ var (
 	httpAddr = flag.String("http", defaultAddr, "HTTP service address")
 	static   = flag.String("static", "", "Path to the static assets")
 	sock     = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	useTCP   = flag.Bool("tcp", false, "Use a tcp socket instead of a Unix socket")
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: This should not be hard-coded, it should be configurable.
Maybe it's possible to use `io.TeeReader`(see [here](https://golang.org/pkg/io/#TeeReader)) or something similar to multiplex? (so support the input from both and output on both)

**Comment by rvolosatovs** [comment]
It would be nice if you separate this in 2 PRs. One for the demo TRC tool and one for the integration tests

File: cmd/trcd/main.go (hunk 3 above)
Comment (ieperen3039) [reply cmd/trcd/main.go]: as answer to why it is not running in a go routine

File: cmd/srrs/main.go (hunk 10 above)
Comment (ieperen3039) [reply cmd/srrs/main.go]: what is hardcoded here? it is a configuration on itself. default is false anyway

Comment (rvolosatovs) [reply cmd/srrs/main.go]: The port

File: cmd/trcd/main.go (hunk 8 above)
Comment (rvolosatovs) [reply cmd/trcd/main.go]: We do want to send things from TRC, we may introduce a way to configure that though

**Comment by rvolosatovs** [comment]
@ieperen3039 please address the comments

File: cmd/trcd/main.go [hunk 11]
```diff
@@ -1,19 +1,22 @@
...
 	"flag"
-	"io"
 	"net"
 	"os"
 	"path/filepath"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/rvolosatovs/turtlitto/pkg/api"
+	"github.com/rvolosatovs/turtlitto/pkg/trcapi"
+	. "github.com/rvolosatovs/turtlitto/pkg/trcapi/trctest"
```
Comment (golangci) [inline cmd/trcd/main.go]: should not use dot imports

File: cmd/trcd/main.go [hunk 12]
```diff
@@ -29,31 +34,94 @@ func init() {
...
+					aWhile := time.Millisecond * time.Duration(rand.Intn(5000))
+					// wait 0-5 seconds, then send an empty state
+					select {
+					case <-time.After(aWhile):
+						if err := trcConn.SendState(&api.State{}); err != nil {
+							log.Error("Random sendState: " + err.Error())
+						}
+						log.Debug("Sent random state")
+					case <-hBeat:
+						trcConn.Ping()
```
Comment (golangci) [inline cmd/trcd/main.go]: Error return value of `trcConn.Ping` is not checked

**Comment by helaan** [comment]
On my box, I need to remove the socket when I interrupt it with Ctrl+C, otherwise the mock trc crashes with 
```
DEBU Listening to Unix socket...                  
panic: listen unix /tmp/trc.sock: bind: address already in use

goroutine 1 [running]:
main.main()
        /home/henk/.go/src/github.com/rvolosatovs/turtlitto/cmd/trcd/main.go:53 +0x788
```
Can this be fixed?

File: cmd/srrs/main.go [hunk 13]
```diff
@@ -52,15 +53,28 @@ func main() {
 	pool := trcapi.NewPool(func() (*trcapi.Conn, func(), error) {
 		logger := log.WithField("path", *sock)
 
-		logger.Debug("Dialing Unix socket...")
-		unixConn, err := net.Dial("unix", *sock)
-		if err != nil {
-			return nil, nil, errors.Wrapf(err, "failed to connect to TRC's unix socket")
+		var socketConn net.Conn
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: `netConn`, not `socketConn`

File: cmd/trcd/main.go [hunk 14]
```diff
@@ -1,19 +1,29 @@
...
+	"time"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/rvolosatovs/turtlitto/pkg/api"
+	"github.com/rvolosatovs/turtlitto/pkg/api/apitest"
+	"github.com/rvolosatovs/turtlitto/pkg/trcapi"
+	"github.com/rvolosatovs/turtlitto/pkg/trcapi/trctest"
 )
 
+// port for connection TRC <-> SRRS when tcp is used
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: `// internalPort represents ...`

File: cmd/trcd/main.go [hunk 15]
```diff
@@ -29,31 +39,96 @@ func init() {
...
-		if _, err := io.Copy(c, os.Stdin); err != nil {
-			log.Printf("Failed to copy input: %s", err)
+
+		if err := trcConn.SendHandshake(&api.Handshake{Version: trcapi.DefaultVersion}); err != nil {
+			panic(err)
+		}
+		log.Debug("Received handshake")
+
+		if err := trcConn.SendState(&api.State{}); err != nil {
+			log.Error("initial state: " + err.Error())
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: Start log statements with a capital

File: cmd/trcd/main.go [hunk 16]
```diff
@@ -29,31 +39,96 @@ func init() {
...
+			go func() {
+				hBeat := time.Tick(3 * time.Second)
+				for {
+					aWhile := time.Millisecond * time.Duration(rand.Intn(5000))
+					// wait 0-5 seconds, then send an empty state
+					select {
+					case <-time.After(aWhile):
+						if err := trcConn.SendState(apitest.RandomState()); err != nil {
+							log.Error("Random sendState error: " + err.Error())
+						}
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: There should be an `else` here, otherwise you print both the error and the `Sent`

**Comment by rvolosatovs** [comment]
@helaan Good point!
We can use https://golang.org/pkg/os/signal/#Notify for it

File: cmd/srrs/main.go [hunk 17]
```diff
@@ -22,10 +22,11 @@ const (
...
 var (
-	debug    = flag.Bool("debug", false, "Debug mode")
-	httpAddr = flag.String("http", defaultAddr, "HTTP service address")
-	static   = flag.String("static", "", "Path to the static assets")
-	sock     = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	debug      = flag.Bool("debug", false, "Debug mode")
+	httpAddr   = flag.String("http", defaultAddr, "HTTP service address")
+	static     = flag.String("static", "", "Path to the static assets")
+	sock       = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	tcpAddress = flag.String("tcpAddr", "", "Service address of tcp socket. TCP will be used instead of a Unix socket when this is set")
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: `tcpSocket` maybe to be consistent with the `socket`?

File: cmd/trcd/main.go [hunk 18]
```diff
@@ -1,19 +1,25 @@
...
+	"github.com/rvolosatovs/turtlitto/pkg/api/apitest"
+	"github.com/rvolosatovs/turtlitto/pkg/trcapi"
+	"github.com/rvolosatovs/turtlitto/pkg/trcapi/trctest"
 )
 
 var (
-	sock = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	debug      = flag.Bool("debug", false, "Debug mode")
+	sock       = flag.String("socket", filepath.Join(os.TempDir(), "trc.sock"), "Path to the unix socket")
+	tcpAddress = flag.String("tcpAddr", "", "Service address of tcp socket. TCP will be used instead of a Unix socket when this is set")
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: `tcpSocket`

File: pkg/api/apitest/apitest.go [hunk 19]
```diff
@@ -51,7 +51,12 @@ func RandomHandshake() *api.Handshake {
 	if err != nil {
 		panic("could not parse version")
 	}
-	return &api.Handshake{ver}
+	b := make([]byte, 10+rand.Intn(10))
+	rand.Read(b)
```
Comment (golangci) [inline pkg/api/apitest/apitest.go]: G404: Use of weak random number generator (math/rand instead of crypto/rand)

File: cmd/trcd/main.go [hunk 20]
```diff
@@ -29,31 +35,98 @@ func init() {
...
+		setStateHandler := func(msg *api.Message) (*api.Message, error) {
+			log.Infof("Received: %s", msg)
+
+			reply, err := trctest.DefaultStateHandler(msg)
+			log.Debugf("Reply with: %s", reply)
+			return reply, err
+		}
+		// ping handler of mock TRC, logs when ping is received
+		pingHandler := func(msg *api.Message) (*api.Message, error) {
+			log.Debug("Received ping")
```
Comment (rvolosatovs) [inline cmd/trcd/main.go]: Log this to `info`?

File: cmd/trcd/main.go [hunk 21]
```diff
@@ -29,31 +35,145 @@ func init() {
...
+					return
+				}
+
+				for {
+					select {
+					case <-time.After(5*time.Second + time.Millisecond*time.Duration(rand.Intn(3000))):
+						if err := trcConn.SendState(apitest.RandomState()); err != nil {
+							log.WithError(err).Error("Failed to send state")
+							return
+						} else {
```
Comment (golangci) [inline cmd/trcd/main.go]: if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)

File: cmd/trcd/main.go [hunk 22]
```diff
@@ -29,31 +35,145 @@ func init() {
...
+							log.Debug("Sent state")
+						}
+
+					case <-time.After(3*time.Second + time.Millisecond*time.Duration(rand.Intn(3000))):
+						if err := trcConn.Ping(); err != nil {
+							log.WithError(err).Error("Failed to send ping")
+							return
+						} else {
```
Comment (golangci) [inline cmd/trcd/main.go]: if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)

**Pull request merged**